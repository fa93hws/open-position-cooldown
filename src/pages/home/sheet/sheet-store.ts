import { makeObservable, observable, action } from 'mobx';

export class SheetStore {
  isRender = false;

  hasTransform = false;

  readonly animationTime = 300;

  constructor() {
    makeObservable(this, {
      isRender: observable.ref,
      hasTransform: observable.ref,
      toggleDisplay: action,
    });
  }

  toggleDisplay(value: boolean) {
    if (this.isRender && !value) {
      // open -> close
      this.hasTransform = false;
      setTimeout(
        action(() => {
          this.isRender = false;
        }),
        this.animationTime,
      );
    } else {
      this.isRender = value;
    }
  }

  onMount() {
    setTimeout(
      action(() => {
        this.hasTransform = this.isRender;
      }),
      50,
    );
  }
}
