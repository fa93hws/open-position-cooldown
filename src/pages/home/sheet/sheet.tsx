import * as React from 'react';
import { createPortal } from 'react-dom';
import { observer } from 'mobx-react';
import { Box, Button } from 'rebass';

import { Plus } from '../../../resources/icons';
import { SheetStore } from './sheet-store';

export const Sheet = (props: {
  onMount: () => void;
  onClose: () => void;
  hasTransform: boolean;
  animationTime: number;
}) => {
  const sx = {
    position: 'absolute',
    transition: `transform ${props.animationTime}ms`,
    transform: 'none',
  };
  if (props.hasTransform) {
    sx.transform = 'translateY(-100%)';
  }
  React.useEffect(() => props.onMount(), []);
  return (
    <Box height="100vh" width="100vw" backgroundColor="white" sx={sx}>
      <Button
        variant="primaryIcon"
        size={['iconSmall', 'iconMedium']}
        sx={{ fill: 'white', transform: 'rotate(45deg)' }}
        p={1}
        onClick={props.onClose}
      >
        <Plus />
      </Button>
    </Box>
  );
};

export function createSheet(): {
  Component: React.ComponentType;
  open(): void;
} {
  const sheetContainer = document.createElement('div');
  document.body.appendChild(sheetContainer);

  const sheetStore = new SheetStore();

  const Component = observer(() => {
    const onMount = () => sheetStore.onMount();
    const onClose = () => sheetStore.toggleDisplay(false);
    return createPortal(
      sheetStore.isRender && (
        <Sheet
          onMount={onMount}
          onClose={onClose}
          hasTransform={sheetStore.hasTransform}
          animationTime={sheetStore.animationTime}
        />
      ),
      sheetContainer,
    );
  });

  return {
    Component,
    open: () => sheetStore.toggleDisplay(true),
  };
}
