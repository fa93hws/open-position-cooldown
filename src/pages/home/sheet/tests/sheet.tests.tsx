import * as React from 'react';

import { Sheet } from '../sheet';

describe('Sheet', () => {
  it('renders it', () => {
    expect(
      <Sheet
        onClose={jest.fn()}
        onMount={jest.fn()}
        hasTransform={false}
        animationTime={300}
      />,
    ).toMatchRenderedSnapshot();
  });
});
