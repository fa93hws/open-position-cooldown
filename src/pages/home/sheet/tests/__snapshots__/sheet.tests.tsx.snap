// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sheet renders it 1`] = `
<div
  class="css-ftf5xp"
>
  <button
    class="css-a6dy74"
  >
    <i />
  </button>
</div>
`;
